process-engine/process_engine_runtime
# ✨ Add Sequelize based migrations

**Changes:**

- Add scripts and configs for sequelize-based migrations, specific for the meta repository.
- These migrations currently support migrating all the sqlite storages from the old specs as implemented by the Skeleton-Electron App to the new Runtime-Specs
- Supports all currently used storages: process_models, flow_node_instances and timers
- Each storage has its own folder and its own migration scripts within the `sequelize/migrations` folder

The migration is done by [Umzug](https://github.com/sequelize/umzug) and can be run by the backend itself. 

No further configs are required.

Currently, migration is executed, when the backend is first launched. To allow the BPMN Studio, or anybody else, to specify a custom path to the Sqlite storages, the startup is no longer executed automatically.
Instead, a function is returned, which accepts the sqlite storage path as a parameter.

To compensate for possible function loss, a `bin/index.js` was added, which starts up the backend without providing a custom sqlite path.

**Issues:**
Closes #31 

**PR:** #30

## How can others test the changes?

- Clear all sqlite databases you might have
- Start the latest Skeleton-Electron app in order to populate some database in the old design
- Fill those databases with data (Especially FlowNodeInstances and ProcessTokens)
- Run the migration script via `npm run migrate-postgres` or `npm run migrate-sqlite`
- See that all data was migrated successfully.

**NOTE ON SQLITE**:
Keep in mind that the Skeleton-Electron App stored its databases in a different location than the runtime!

## PR-Checklist

Please check the boxes in this list after submitting your PR:


## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>
